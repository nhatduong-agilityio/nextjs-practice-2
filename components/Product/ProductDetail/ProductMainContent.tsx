import { Box, Flex, Heading, Text } from '@chakra-ui/react'
import React, { Fragment, memo } from 'react'
import { v4 as uuidv4 } from 'uuid'
import StarIcon from '@components/Icons/StarIcon'

interface IProps {
  rating: number
  price: number
  discount: number
  sellCount: number
}

const ProductMainContent = memo(({ rating, discount, price, sellCount }: IProps) => {
  // valid sales percent
  const validDiscount = discount > 0 && discount <= 100

  // calculator price
  const salePrice = price - (price * discount) / 100

  return (
    <Box w='full' style={{ margin: 0 }}>
      <Box display='flex' pt='15px' alignItems='center'>
        {Array(5)
          .fill('')
          .map((_, index) => (
            <Fragment key={uuidv4()}>
              <StarIcon
                width={['15px', '18px', '18px']}
                height={['15px', '18px', '18px']}
                viewBox='0 0 15 15'
                fill={index < rating ? '#F9BB1D' : 'gray.300'}
                style={{ marginRight: '1px' }}
              />
            </Fragment>
          ))}
        <Text ml='10px' size={['16', '18']} variant='grayScheme'>
          ({sellCount} Sell)
        </Text>
      </Box>
      <Flex flexDir='row' alignItems='center' w='full' pt='15px' gap={['5px', '10px']}>
        <Heading mr='10px' fontSize={['20px', '45px']}>
          ${salePrice}
        </Heading>
        {validDiscount && (
          <>
            <Text as='del' size={['18', '40']} mr='10px'>
              ${price}
            </Text>
            <Text size={['18', '40']} variant='redScheme'>
              -{discount}%
            </Text>
          </>
        )}
      </Flex>
    </Box>
  )
})

ProductMainContent.displayName = 'ProductProductMainContent'

export default ProductMainContent
