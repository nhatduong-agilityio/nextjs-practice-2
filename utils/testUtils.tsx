import { ChakraProvider } from '@chakra-ui/react'
import { AuthProvider } from '@store/providers/AuthProvider'
import { LoadingProvider } from '@store/providers/LoadingProvider'
import type { RenderOptions, RenderResult } from '@testing-library/react'
import { render } from '@testing-library/react'
import { customTheme } from '@themes/index'
import type { ReactElement } from 'react'

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <ChakraProvider theme={customTheme}>
      <LoadingProvider>
        <AuthProvider>{children}</AuthProvider>
      </LoadingProvider>
    </ChakraProvider>
  )
}

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>): RenderResult =>
  render(ui, { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }
