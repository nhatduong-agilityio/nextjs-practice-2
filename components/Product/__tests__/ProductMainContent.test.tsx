// Utils
import { render } from '@utils/testUtils'

// Components
import { ProductsMock } from '@mock/dataMock'
import type { TProduct } from '@matched-types/product'
import ProductMainContent from '../ProductDetail/ProductMainContent'

describe('ProductMainContent render', () => {
  const productMock = ProductsMock[0] ? ProductsMock[0] : ({} as TProduct)

  it('Should ProductMainContent match Snapshot', () => {
    const { container } = render(
      <ProductMainContent
        rating={productMock.rating}
        price={productMock.price}
        discount={productMock.discount}
        sellCount={productMock.sellCount}
      />,
    )

    expect(container).toMatchSnapshot()
  })
})
