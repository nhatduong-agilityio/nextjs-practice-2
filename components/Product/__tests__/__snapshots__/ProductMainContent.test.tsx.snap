// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductMainContent render Should ProductMainContent match Snapshot 1`] = `
<div>
  <div
    class="css-1ki54i"
    style="margin: 0px;"
  >
    <div
      class="css-ow32u0"
    >
      <svg
        class="chakra-icon css-tlkdll"
        focusable="false"
        style="margin-right: 1px;"
        viewBox="0 0 15 15"
      >
        <path
          d="M7.5 10.7938L11.3625 13.125L10.3375 8.73125L13.75 5.775L9.25625 5.39375L7.5 1.25L5.74375 5.39375L1.25 5.775L4.6625 8.73125L3.6375 13.125L7.5 10.7938Z"
        />
      </svg>
      <svg
        class="chakra-icon css-tlkdll"
        focusable="false"
        style="margin-right: 1px;"
        viewBox="0 0 15 15"
      >
        <path
          d="M7.5 10.7938L11.3625 13.125L10.3375 8.73125L13.75 5.775L9.25625 5.39375L7.5 1.25L5.74375 5.39375L1.25 5.775L4.6625 8.73125L3.6375 13.125L7.5 10.7938Z"
        />
      </svg>
      <svg
        class="chakra-icon css-tlkdll"
        focusable="false"
        style="margin-right: 1px;"
        viewBox="0 0 15 15"
      >
        <path
          d="M7.5 10.7938L11.3625 13.125L10.3375 8.73125L13.75 5.775L9.25625 5.39375L7.5 1.25L5.74375 5.39375L1.25 5.775L4.6625 8.73125L3.6375 13.125L7.5 10.7938Z"
        />
      </svg>
      <svg
        class="chakra-icon css-tlkdll"
        focusable="false"
        style="margin-right: 1px;"
        viewBox="0 0 15 15"
      >
        <path
          d="M7.5 10.7938L11.3625 13.125L10.3375 8.73125L13.75 5.775L9.25625 5.39375L7.5 1.25L5.74375 5.39375L1.25 5.775L4.6625 8.73125L3.6375 13.125L7.5 10.7938Z"
        />
      </svg>
      <svg
        class="chakra-icon css-tlkdll"
        focusable="false"
        style="margin-right: 1px;"
        viewBox="0 0 15 15"
      >
        <path
          d="M7.5 10.7938L11.3625 13.125L10.3375 8.73125L13.75 5.775L9.25625 5.39375L7.5 1.25L5.74375 5.39375L1.25 5.775L4.6625 8.73125L3.6375 13.125L7.5 10.7938Z"
        />
      </svg>
      <p
        class="chakra-text css-1kbeja5"
      >
        (
        400
         Sell)
      </p>
    </div>
    <div
      class="css-1rqbrwb"
    >
      <h2
        class="chakra-heading css-1naee9b"
      >
        $
        120
      </h2>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
